File: notificare-inbox/android/src/main/java/re/notifica/inbox/react_native/NotificareInboxModule.kt
package re.notifica.inbox.react_native

import android.os.Handler
import android.os.Looper
import androidx.lifecycle.Observer
import com.facebook.react.bridge.*
import re.notifica.NotificareCallback
import re.notifica.NotificareLogger
import re.notifica.inbox.NotificareInbox
import re.notifica.inbox.models.NotificareInboxItem
import re.notifica.models.NotificareNotification
import java.util.*

class NotificareInboxModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

  private val itemsObserver = Observer<SortedSet<NotificareInboxItem>> { items ->
    if (items == null) return@Observer

    try {
      val payload = Arguments.createArray()
      items.forEach {
        payload.pushMap(it.toJson().toReactMap())
      }

      NotificareInboxModuleEventManager.dispatchEvent("inbox_updated", payload)
    } catch (e: Exception) {
      NotificareLogger.error("Failed to emit the inbox_updated event.", e)
    }
  }

  private val badgeObserver = Observer<Int> { badge ->
    if (badge == null) return@Observer

    NotificareInboxModuleEventManager.dispatchEvent("badge_updated", badge)
  }

  override fun getName(): String = "NotificareInboxModule"

  override fun initialize() {
    super.initialize()

    NotificareInboxModuleEventManager.setup(reactApplicationContext)

    onMainThread {
      NotificareInbox.observableItems.observeForever(itemsObserver)
      NotificareInbox.observableBadge.observeForever(badgeObserver)
    }
  }

  override fun onCatalystInstanceDestroy() {
    super.onCatalystInstanceDestroy()

    onMainThread {
      NotificareInbox.observableItems.removeObserver(itemsObserver)
      NotificareInbox.observableBadge.removeObserver(badgeObserver)
    }
  }

  @ReactMethod
  fun getItems(promise: Promise) {
    try {
      val payload = Arguments.createArray()
      NotificareInbox.items.forEach {
        payload.pushMap(it.toJson().toReactMap())
      }

      promise.resolve(payload)
    } catch (e: Exception) {
      promise.reject(DEFAULT_ERROR_CODE, e)
    }
  }

  @ReactMethod
  fun getBadge(promise: Promise) {
    promise.resolve(NotificareInbox.badge)
  }

  @ReactMethod
  fun refresh(promise: Promise) {
    NotificareInbox.refresh()
    promise.resolve(null)
  }

  @ReactMethod
  fun open(data: ReadableMap, promise: Promise) {
    val item: NotificareInboxItem

    try {
      item = NotificareInboxItem.fromJson(data.toJson())
    } catch (e: Exception) {
      promise.reject(DEFAULT_ERROR_CODE, e)
      return
    }

    NotificareInbox.open(item, object : NotificareCallback<NotificareNotification> {
      override fun onSuccess(result: NotificareNotification) {
        try {
          promise.resolve(result.toJson().toReactMap())
        } catch (e: Exception) {
          promise.reject(DEFAULT_ERROR_CODE, e)
        }
      }

      override fun onFailure(e: Exception) {
        promise.reject(DEFAULT_ERROR_CODE, e)
      }
    })
  }

  @ReactMethod
  fun markAsRead(data: ReadableMap, promise: Promise) {
    val item: NotificareInboxItem

    try {
      item = NotificareInboxItem.fromJson(data.toJson())
    } catch (e: Exception) {
      promise.reject(DEFAULT_ERROR_CODE, e)
      return
    }

    NotificareInbox.markAsRead(item, object : NotificareCallback<Unit> {
      override fun onSuccess(result: Unit) {
        promise.resolve(null)
      }

      override fun onFailure(e: Exception) {
        promise.reject(DEFAULT_ERROR_CODE, e)
      }
    })
  }

  @ReactMethod
  fun markAllAsRead(promise: Promise) {
    NotificareInbox.markAllAsRead(object : NotificareCallback<Unit> {
      override fun onSuccess(result: Unit) {
        promise.resolve(null)
      }

      override fun onFailure(e: Exception) {
        promise.reject(DEFAULT_ERROR_CODE, e)
      }
    })
  }

  @ReactMethod
  fun remove(data: ReadableMap, promise: Promise) {
    val item: NotificareInboxItem

    try {
      item = NotificareInboxItem.fromJson(data.toJson())
    } catch (e: Exception) {
      promise.reject(DEFAULT_ERROR_CODE, e)
      return
    }

    NotificareInbox.remove(item, object : NotificareCallback<Unit> {
      override fun onSuccess(result: Unit) {
        promise.resolve(null)
      }

      override fun onFailure(e: Exception) {
        promise.reject(DEFAULT_ERROR_CODE, e)
      }
    })
  }

  @ReactMethod
  fun clear(promise: Promise) {
    NotificareInbox.clear(object : NotificareCallback<Unit> {
      override fun onSuccess(result: Unit) {
        promise.resolve(null)
      }

      override fun onFailure(e: Exception) {
        promise.reject(DEFAULT_ERROR_CODE, e)
      }
    })
  }

  companion object {
    internal const val DEFAULT_ERROR_CODE = "notificare_error"

    internal fun onMainThread(action: () -> Unit) = Handler(Looper.getMainLooper()).post(action)
  }
}

File: notificare-inbox/ios/NotificareInboxModule.swift
import React
import NotificareKit
import NotificareInboxKit

fileprivate let DEFAULT_ERROR_CODE = "notificare_error"

@objc(NotificareInboxModule)
class NotificareInboxModule: RCTEventEmitter {

    private var hasListeners = false
    private var eventQueue = [(name: String, payload: Any?)]()
    
    override init() {
        super.init()
        
        NotificareInbox.shared.delegate = self
    }
    
    override class func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    override func startObserving() {
        hasListeners = true
        
        if !eventQueue.isEmpty {
            NotificareLogger.debug("Processing event queue with \(eventQueue.count) items.")
            eventQueue.forEach { sendEvent(withName: $0.name, body: $0.payload)}
            eventQueue.removeAll()
        }
    }
    
    override func stopObserving() {
        hasListeners = false
    }
    
    override func supportedEvents() -> [String] {
        return [
            "inbox_updated",
            "badge_updated",
        ]
    }
    
    private func dispatchEvent(_ name: String, payload: Any?) {
        if hasListeners {
            sendEvent(withName: name, body: payload)
        } else {
            eventQueue.append((name: name, payload: payload))
        }
    }
 
    // MARK: - Notificare Inbox
    
    @objc
    func getItems(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        do {
            let payload = try NotificareInbox.shared.items.map { try $0.toJson() }
            resolve(payload)
        } catch {
            reject(DEFAULT_ERROR_CODE, error.localizedDescription, nil)
        }
    }
    
    @objc
    func getBadge(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        resolve(NotificareInbox.shared.badge)
    }
    
    @objc
    func open(_ data: [String: Any], resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        let item: NotificareInboxItem
        
        do {
            item = try NotificareInboxItem.fromJson(json: data)
        } catch {
            reject(DEFAULT_ERROR_CODE, error.localizedDescription, nil)
            return
        }
        
        NotificareInbox.shared.open(item) { result in
            switch result {
            case let .success(notification):
                do {
                    let payload = try notification.toJson()
                    resolve(payload)
                } catch {
                    reject(DEFAULT_ERROR_CODE, error.localizedDescription, nil)
                }
            case let .failure(error):
                reject(DEFAULT_ERROR_CODE, error.localizedDescription, nil)
            }
        }
    }
    
    @objc
    func markAsRead(_ data: [String: Any], resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        let item: NotificareInboxItem
        
        do {
            item = try NotificareInboxItem.fromJson(json: data)
        } catch {
            reject(DEFAULT_ERROR_CODE, error.localizedDescription, nil)
            return
        }
        
        NotificareInbox.shared.markAsRead(item) { result in
            switch result {
            case .success:
                resolve(nil)
            case let .failure(error):
                reject(DEFAULT_ERROR_CODE, error.localizedDescription, nil)
            }
        }
    }
    
    @objc
    func markAllAsRead(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        NotificareInbox.shared.markAllAsRead { result in
            switch result {
            case .success:
                resolve(nil)
            case let .failure(error):
                reject(DEFAULT_ERROR_CODE, error.localizedDescription, nil)
            }
        }
    }
    
    @objc
    func remove(_ data: [String: Any], resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        let item: NotificareInboxItem
        
        do {
            item = try NotificareInboxItem.fromJson(json: data)
        } catch {
            reject(DEFAULT_ERROR_CODE, error.localizedDescription, nil)
            return
        }
        
        NotificareInbox.shared.remove(item) { result in
            switch result {
            case .success:
                resolve(nil)
            case let .failure(error):
                reject(DEFAULT_ERROR_CODE, error.localizedDescription, nil)
            }
        }
    }
    
    @objc
    func clear(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        NotificareInbox.shared.clear { result in
            switch result {
            case .success:
                resolve(nil)
            case let .failure(error):
                reject(DEFAULT_ERROR_CODE, error.localizedDescription, nil)
            }
        }
    }
}

extension NotificareInboxModule: NotificareInboxDelegate {
    func notificare(_ notificareInbox: NotificareInbox, didUpdateInbox items: [NotificareInboxItem]) {
        do {
            dispatchEvent("inbox_updated", payload: try items.map { try $0.toJson() })
        } catch {
            NotificareLogger.error("Failed to emit the inbox_updated event.\n\(error)")
        }
    }
    
    func notificare(_ notificareInbox: NotificareInbox, didUpdateBadge badge: Int) {
        dispatchEvent("badge_updated", payload: badge)
    }
}
